rasa_core/training/structures.py

```diff
@@ -274,24 +284,89 @@ def with_cycles_removed(self):
                     updated.start_checkpoints.extend(additional_ends)
                     story_steps[k] = updated
 
             if needs_connector:
-                start.end_checkpoints.append(Checkpoint(connector_cid))
+                start.end_checkpoints.append(Checkpoint(connector_cp_name))
+
+        self._remove_unused_generated_cps(story_steps,
+                                          all_overlapping_cps,
+                                          story_end_checkpoints)
 
-        return StoryGraph(story_steps.values(),
+        return StoryGraph(list(story_steps.values()),
                           story_end_checkpoints)
 
+    @staticmethod
+    def _checkpoint_difference(cps, cp_to_ignore):
+        return [cp for cp in cps if cp.name not in cp_to_ignore]
+
+    def _remove_unused_generated_cps(self, story_steps, overlapping_cps,
+                                     story_end_checkpoints):
+        # the process above may generate unused start checkpoints
+        # we need to find them and remove them
+        # also there might be generated unused end checkpoints
+        unused_cps = self._unused_checkpoints(story_steps.values(),
+                                              story_end_checkpoints)
+
+        unused_overlapping_cps = unused_cps.intersection(overlapping_cps)
+
+        unused_genr_cps = {cp_name
+                           for cp_name in unused_cps
+                           if cp_name.startswith(GENERATED_CHECKPOINT_PREFIX)}
+
+        for k, step in story_steps.items():
+            # changed all ends
+            updated = step.create_copy(use_new_id=False)
+            updated.start_checkpoints = self._checkpoint_difference(
+                    updated.start_checkpoints, unused_overlapping_cps)
+
+            # remove generated unused end checkpoints
+            updated.end_checkpoints = self._checkpoint_difference(
+                    updated.end_checkpoints, unused_genr_cps)
+
+            story_steps[k] = updated
+
+    @staticmethod
+    def _is_checkpoint_in_list(checkpoint_name, conditions, cps):
+        for cp in cps:
+            if checkpoint_name == cp.name and conditions == cp.conditions:
+                return True
+        return False
+
+    @staticmethod
+    def _unused_checkpoints(story_steps, story_end_checkpoints):
+        """Find all end checkpoints."""
+
+        collected_start = {STORY_END, STORY_START}
+        collected_end = {STORY_END, STORY_START}
+
+        for step in story_steps:
+            for start in step.start_checkpoints:
+                collected_start.add(start.name)
+            for end in step.end_checkpoints:
+                if end.name in story_end_checkpoints.keys():
+                    end_name = story_end_checkpoints[end.name]
+                else:
+                    end_name = end.name
+                collected_end.add(end_name)
+
+        return collected_end.union(collected_start)
```

Comment:
I don't see how this function computes unused checkpoints ❓

Reply:
well, you changed it to `union` :stuck_out_tongue_winking_eye:, it should be `^`=`XOR`

Reply:
very good point :D this makes a lot more sense, changed it to `symmetric_difference` which is XOR.